tests/simulation/simul_noisify_2.py

```diff
@@ -0,0 +1,138 @@
+"""
+This is file contains a sample main function.
+You can build your own, to simulate and test your data.
+
+It is not a TEST file, it is a sample file to run the algorithm.
+"""
+
+import pdb
+from typing import OrderedDict
+from xmlrpc.client import boolean
+from src.dedupe_gazetteer import *
+from src.dedupe_gazetteer_utils import *
+
+
+import logging
+from datetime import datetime
+import optparse
+import _io, os
+import unittest
+import dedupe
+import itertools
+from collections import defaultdict
+from settings import *
+
+
+
+def train(variables):
+    training_element = TrainingElement(op_training_file, variables)
+    """
+    Skipt training with you have already trained otherwhise it will delete your trained and settings file,
+    and train again.
+    """
+    tp = TrainingProcess(ip_canonical_file, op_settings_file, op_training_file, training_element)
+    print(f"Number of records from canonical data (pesquisadores unicos): {len(tp.canonical_d)}")
+
+    sample_size = 1000
+    step_time = datetime.now()
+    tp.training(ip_messy_training_file, ip_messy_validation_file, sample_size = 1000)
+    end_time = datetime.now()
+    print(f"Training Time: {end_time - step_time} \n")
+
+
+def predict(noise):
+    start_time = datetime.now()
+    step_time = 0
+    end_time = 0
+
+    ### Prediction
+    step_time = datetime.now()
+    model_evaluation = ModelEvaluation(ip_canonical_file, ip_messy_test_file, op_false_positives_file, op_false_negatives_file)
+    noisify = Noisify(noise[0], noise[1])
+    tt = TrainingTest(ip_canonical_file, ip_messy_test_file, op_settings_file, op_matches_found_file, model_evaluation, noisify)
+    tt.noisify()
+    tt.cluster_data()
+    end_time = datetime.now()
+    print(f"Prediction Time and File Recording Time: {end_time - step_time} \n")
+    ###
+
+    ### Performance of the algorithm
+    step_time = datetime.now()
+    tt.evaluate_model()
+    end_time = datetime.now()
+    print(f"Model Evaluation Time and Files Recording Time: {end_time - step_time} \n")
+    ###
+
+    print(f'Total Processing Time: {end_time - start_time}')
+
+
+
+
+def build_variables():
+    primeiro_nome = ['Exact', 'String']
+    ultimo_sobrenome = ['Exact', 'String']
+    return list(itertools.product(primeiro_nome, ultimo_sobrenome))
+
+def build_noise():
+    ruido = [1,2]
+    abs_percent = [True, False]
+    return list(itertools.product(ruido, abs_percent))
+
+def build_simulation():
+    simulations = []
+    variables = build_variables()
```

Comment:
Suggestion:
It may be preferable to rename the variable 'variables' because it can be misinterpreted when you change it, on line 90, to set it with the types of variables that Dedupe will pay attention to.